src/App.js

```diff
@@ -23,49 +25,107 @@ const generateSquares = () => {
   }
 
   return squares;
-}
+};
 
 const App = () => {
-
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
+
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState(null);
+
+  const onClickCallback = (id) => {
+    setSquares((squares) =>{
+      let newBoard = squares.map((square)=>{
+        for (let property of square){
+          if (property.id === id && property.value === ''){
+            if (currentPlayer === PLAYER_1){
+              property.value = PLAYER_1;
+            } else if (currentPlayer === PLAYER_2){
+              property.value = PLAYER_2;
+            }
+          }
```

Comment:
If you rename the map param to row, it's more clear that what this code does is iterate over each square data (called property here) from the original state. And when modifying each square data, both the old state and the new state refer to that square data, so both end up being modified. So there's cross contamination of the old and new states. We can tell this is happening, because the checkForWinner call on like 52 is working. It refers to the current squares state, and that shouldn't get updated until the next render pass. But since it's working for it to be called in the click handler (which still runs using the data from the previous render) then previous state must be getting updated. We really need to do a doubly-nested map, so that both layers of lists are duplicated, and then also spread the square data that gets updated so that only the new list has the new data in it.